Sming/SmingCore/Wire.cpp

```diff
@@ -1,145 +1,249 @@
 }
 
-size_t TwoWire::write(uint8_t data)
-{
-    if(txLen >= BUFFER_LENGTH || txLen == -1)
-    {
-      txLen = -1; // Overflow :(
+size_t TwoWire::write(uint8_t data){
+  if(transmitting){
+    if(txBufferLength >= BUFFER_LENGTH){
+      setWriteError();
       return 0;
     }
+    txBuffer[txBufferIndex] = data;
+    ++txBufferIndex;
+    txBufferLength = txBufferIndex;
+  } else {
+    // i2c_slave_transmit(&data, 1);
+  }
+  return 1;
+}
 
-    txBuf[txLen++] = data;
-    return 1;
+size_t TwoWire::write(const uint8_t *data, size_t quantity){
+  if(transmitting){
+    for(size_t i = 0; i < quantity; ++i){
+      if(!write(data[i])) return i;
+    }
+  }else{
+    // i2c_slave_transmit(data, quantity);
+  }
+  return quantity;
 }
 
-int TwoWire::available()
-{
-	return rxLen - rxPos;
+int TwoWire::available(void){
+  int result = rxBufferLength - rxBufferIndex;
+  return result;
 }
 
-int TwoWire::read()
-{
-	if (rxLen > rxPos)
-		return rxBuf[rxPos++];
-	else
-		return -1;
+int TwoWire::read(void){
+  int value = -1;
+  if(rxBufferIndex < rxBufferLength){
+    value = rxBuffer[rxBufferIndex];
+    ++rxBufferIndex;
+  }
+  return value;
 }
 
-int TwoWire::peek()
-{
-	return rxBuf[rxPos];
+int TwoWire::peek(void){
+  int value = -1;
+  if(rxBufferIndex < rxBufferLength){
+    value = rxBuffer[rxBufferIndex];
+  }
+  return value;
 }
 
-void TwoWire::flush()
-{
+void TwoWire::flush(void){
+  rxBufferIndex = 0;
+  rxBufferLength = 0;
+  txBufferIndex = 0;
+  txBufferLength = 0;
 }
 
-size_t TwoWire::write(const uint8_t *data, size_t quantity)
+void TwoWire::onReceiveService(uint8_t* inBytes, int numBytes)
 {
-	for(size_t i = 0; i < quantity; i++)
-		write(data[i]);
+  // don't bother if user hasn't registered a callback
+  // if(!user_onReceive){
+  //   return;
+  // }
+  // // don't bother if rx buffer is in use by a master requestFrom() op
+  // // i know this drops data, but it allows for slight stupidity
+  // // meaning, they may not have read all the master requestFrom() data yet
+  // if(rxBufferIndex < rxBufferLength){
+  //   return;
+  // }
+  // // copy twi rx buffer into local read buffer
+  // // this enables new reads to happen in parallel
+  // for(uint8_t i = 0; i < numBytes; ++i){
+  //   rxBuffer[i] = inBytes[i];
+  // }
+  // // set rx iterator vars
+  // rxBufferIndex = 0;
+  // rxBufferLength = numBytes;
+  // // alert user program
+  // user_onReceive(numBytes);
+}
 
-	return quantity;
+void TwoWire::onRequestService(void){
+  // // don't bother if user hasn't registered a callback
+  // if(!user_onRequest){
+  //   return;
+  // }
+  // // reset tx buffer iterator vars
+  // // !!! this will kill any pending pre-master sendTo() activity
+  // txBufferIndex = 0;
+  // txBufferLength = 0;
+  // // alert user program
+  // user_onRequest();
 }
 
-TwoWire Wire = TwoWire(I2C_DEFAULT_SCL_PIN, I2C_DEFAULT_SDA_PIN);
+void TwoWire::onReceive( void (*function)(int) ){
+  //user_onReceive = function;
```

Comment:
This is PORT of the library with as minimum changes as possible. May be we should import it as is, try how it works and then add more useful stuff to it as delegates etc??